fn-deps for lambdas

Cyrik opened the pull request:
Fixes #51 

Uses Clojures internal class cache to also check anonymous function dependencies.

bbatsov commented:
Small note - we can't really depend on `tools.reader`, as this might create conflicts with user applications. Orchard has to have no runtime deps.

Cyrik commented:
> Small note - we can't really depend on tools.reader, as this might create conflicts with user applications.

ah thanks, I completely forgot I was still depending on that. I'll rewrite that part to use clojure/read.

Do you have any thoughts on the idea of using the compiler to find the dependencies?

bbatsov commented:
The idea itself seems reasonable to me.

Cyrik commented:
I've got a new solution that doesn't need to recompile the code and is a lot faster.
The problem with the new solution is that old classes stay around in clojures cache for some time. I haven't found a way yet to find out which old classes aren't in use anymore.
But the new solution is already a lot better than the old one.
Do you have any ideas on how to fix the final problem?
Also would you want to keep the recompile version as an extra possibility if the actual bytecode is needed for something else?

vemv commented:
> The problem with the new solution is that old classes stay around in clojures cache for some time. 

Can you explain this as a PR comment (self-review)? That way we can see the code that the problem in question relates to.

Cyrik commented on src/orchard/xref.clj:
```diff
@@ -13,21 +32,87 @@
+  (when-let [^clojure.lang.AFn v (as-val s)]
+    (let [f-class-name (-> v .getClass .getName)
+          field (->> clojure.lang.DynamicClassLoader .getDeclaredFields second)
+          classes (into {} (.get field clojure.lang.DynamicClassLoader))
```
this holds all functions compiled by clojure:
{...
"dev$dummy"
   #object
    [java.lang.ref.SoftReference 0x5081395f
     "java.lang.ref.SoftReference@5081395f"]
...
}
where lambdas are named "dev$dummy$fn_1234"

the filter uses that fact to just keep the function + lambda classes.

The problem is, when you recompile a function, new lambda classes get generated and the old ones stay around.

vemv replied:
>  the old ones stay around

As in, garbage to be collected?

What causes that retention? This is a `defn` (and not e.g. a `def`), so why invoking a defn would retain memory references?

Unless there's global state somewhere, probably your solution is good - it's just minor details like `*1` or other repl state that might retain objects for longer than expected.

Cyrik replied:
I probably explained it wrong.
The problem is not the memory, the problem is that the Clojure classloader keeps the old classes in its own cache as java.lang.ref.SoftReference and only deletes that if you try to access the reference after it has been cleaned up.
So if I do the following in my code:
```clojure
(defn dummy []
  (map #(println %) (range 10)))
(defn dummy []
  (map #(print %) (range 10)))
;;o r just send both versions to the repl by changing the first
```
classes will now contain "dev$dummy$fn_123" and "dev$dummy$fn_124" where 123 references println and 124 references print.
I haven't found an easy way to get rid of "dev$dummy$fn_123" yet.

vemv replied:
Perhaps it would be solved if WeakReference was used instead.

However, as mentioned personally I do not condone classloader hacking :)

Cyrik replied:
This solution does not overwrite the clojure classLoader, so I can't switch the reference type. This just accesses the regular clojure loader through reflection.
I probably shouldn't have kept two separate solutions in the same file, I can see how that is confusing, sorry.

vemv replied:
> I haven't found an easy way to get rid of "dev$dummy$fn_123" yet.

Does the PR still have this problem?

It might be acceptable to have that problem, if so at least leave a comment explaining what's going on

Cyrik replied:
Yes, the PR still has the problem. I'll play around with it some more to see if I can find a solution. If I can't find one I'll add a comment describing the issue and open a new ticket once the PR is merged.

vemv commented on src/orchard/xref.clj:
```diff
@@ -13,21 +32,87 @@
+(defn fn-deps [s]
+  (when-let [^clojure.lang.AFn v (as-val s)]
+    (let [f-class-name (-> v .getClass .getName)
+          field (->> clojure.lang.DynamicClassLoader .getDeclaredFields second)
```
can you please get the field by name? so that we can see what field it is

Cyrik replied:
that's a problem I'm still having, if I change that line to (.getDeclaredField clojure.lang.DynamicClassLoader "classCache") which is basically the same thing, I get an IllegalAccessException on the next line.
The same thing happens if I type hint field to ^java.lang.reflect.Field

vemv replied:
At least `field` could be named more expressively

vemv commented on src/orchard/xref.clj:
```diff
@@ -13,21 +32,87 @@
+(defn- fn-name [^java.lang.Class f]
+  (-> f .getName repl/demunge symbol))
+
+(defonce classbytes (atom {}))
```
We shouldn't have global state, sounds like a recipe for memory leaks

vemv commented on src/orchard/xref.clj:
```diff
@@ -13,21 +32,87 @@
+
+(defonce classbytes (atom {}))
+
+(defn- recompile [ns-sym form]
```
Why is "recompilation" needed? This would be a great `;; comment` to add.

Cyrik replied:
You are right, I should have explained it in the code.
The idea is to recompile the code and overwrite the classLoader just during that recompilation so that we can get at the bytecode. Sadly there is no other way to do that since the JVM just throws all of that away after loading.
This "should" not interfere with normal execution since the new classes are never used and the proxied loader is thrown away after use.
But hopefully, the recompile solution won't be needed if I can fix the final problem with the second solution.
We might want to keep the code around though since the bytecode can be used for other introspection at a later point.

vemv commented on src/orchard/xref.clj:
```diff
@@ -13,21 +32,87 @@
+(defn- recompile [ns-sym form]
+  (push-thread-bindings
+   {clojure.lang.Compiler/LOADER
+    (proxy [clojure.lang.DynamicClassLoader] [@clojure.lang.Compiler/LOADER]
```
Just as a word of caution, hacking with classloaders as a long history of giving us problems, in various tools over the years.

If you have the chance, please avoid these as they can compose badly with other features.

Clojure internals are best kept internal :)

vemv commented:
Part of the context is https://lukas-domagala.de/blog/clojure-analysis-and-introspection.html which says:

> [clj-kondo] works great but has two new problems:
> * it's kind of slow (40 seconds for that project, including jars)
> * does not work for running processes where you don't have the code (remote repls)

Personally I like the middle ground of using clj-kondo but spawn from the same JVM that hosts a given repl.

That way the classpath is always up to date, and one doesn't need external binaries.

clj-kondo itself uses a persistent cache so I don't believe saying it takes 40s to analyze a project is quite representative of day-to-day UX.

As you, I much favor runtime-powered solutions where they make sense. Rich Hickey's hack is a hack, it might work for him because he authored the Clojure compiler.

The compiler's internals are generally not for us to rely on.

So clj-kondo is a fine solution, just as tools.analyzer is (which is very much runtime-powered, so survives any dynamism).

My take would be that unless this PR becomes substantially more minimalistic, it would be not be worth the hassle.

One can always use his own tooling on a personal stack - which I do all the time! Later one can propose PRs as those solutions become more time-proven.

Cyrik commented:
Thank you for your comments @vemv
I completely understand your position. Clj-kondo is great, it's just a little slow for my use case, which it was not meant to handle.

I've removed the recompile part of the code and commented on the weird IllegalAccess workaround. It's completely fine if you don't want to merge this. This way whoever stumbles over this will have a clean thing to build upon.

vemv commented on src/orchard/xref.clj:
```diff
@@ -13,21 +15,40 @@
+    (let [f-class-name (-> v .getClass .getName)
+          ;; breaks when called with
+          ;; (.getDeclaredField clojure.lang.DynamicClassLoader "classCache")
+          classCache (->> clojure.lang.DynamicClassLoader .getDeclaredFields second)
```
Maybe you can `(->> clojure.lang.DynamicClassLoader .getDeclaredFields (filter ...) first)`? Since `second` seems non-deterministic.

Also, the work can be done in a `def` to not repeat the work on each invocation.

vemv commented on src/orchard/xref.clj:
```diff
@@ -13,21 +15,40 @@
+    (let [f-class-name (-> v .getClass .getName)
+          ;; breaks when called with
+          ;; (.getDeclaredField clojure.lang.DynamicClassLoader "classCache")
+          classCache (->> clojure.lang.DynamicClassLoader .getDeclaredFields second)
```
Why not type-hint this? Seems better than skipping Eastwood.

Cyrik replied:
If I type-hint this I get an IllegalAccess Exception, I'm guessing it's the same problem as the direct access.
I'll have a look at the bytecode produced by both versions to hopefully find out why one works and the other doesn't, but it's a lower priority than the other things.

vemv commented on src/orchard/xref.clj:
```diff
@@ -13,21 +15,40 @@
+          ;; (.getDeclaredField clojure.lang.DynamicClassLoader "classCache")
+          classCache (->> clojure.lang.DynamicClassLoader .getDeclaredFields second)
+          classes (into {} (.get classCache clojure.lang.DynamicClassLoader))
+          filtered-classes (->> classes
+                                (filter (fn [[k _v]] (clojure.string/includes? k f-class-name)))
+                                (map (fn [[_k v]] (.get ^java.lang.ref.Reference v))))
+          deps (set (mapcat fn-deps-class filtered-classes))]
```
Probably using transducers would improve performance, there can be quite a lot of classes in a given JVM session.

vemv reviewed (commented):
The PR looks more minimalistic now so we can proceed, thanks!

bbatsov commented:
> I've removed the recompile part of the code and commented on the weird IllegalAccess workaround. It's completely fine if you don't want to merge this. This way whoever stumbles over this will have a clean thing to build upon.

Well, our built-in find-usages is not going away, so I'm in favor of improving it. I gets @vemv's position but I think it's fine to keep improving what we have given the any alternative will require deps and extra work that no one has been willing to do for years. :-) 

I also think that Clojure's compiler API has been super stable, so the concerns about whether it's a good idea to use its API are probably overdone. Let's drive this PR to the finish line.

Cyrik commented on src/orchard/xref.clj:
```diff
@@ -13,21 +15,47 @@
+(def ^java.lang.reflect.Field classCache
+  (let [classCache* (.getDeclaredField clojure.lang.DynamicClassLoader "classCache")]
+    (.setAccessible classCache* true)
+    (.get classCache* clojure.lang.DynamicClassLoader)))
```
I've managed to fix the IllegalAccess issue. This was something I tried before, not sure why it's working now, but I'm happy.

bbatsov replied:
Just add some comments explaining this for posterity's sake.

Cyrik commented on src/orchard/xref.clj:
```diff
@@ -13,21 +15,47 @@
+                                   (.get f (fn-name v)))
                               nil)))))))
 
+(def ^java.lang.reflect.Field classCache
```
the type-hint is wrong, I'll fix that in the next commit

Cyrik commented:
I'm still trying to fix the "old Lamdba" reference problem but otherwise this is done.
There is a hacky fix where you force an OOM Exception on purpose, but it doesn't always work and feels very dirty. Hopefully I'll find another way, or I'll leave it like this.

vemv commented:
Kudos for the iteration!

There's `System/gc`, have you tried that? It might make sense to invoke it whenever the user invokes the functionality (as opposed to on a schedule or something similarly invasive).

WIth WeakReference (vs. SoftReference), GCs are even more effective, but I realise we cannot fork Clojure.

Cyrik commented:
gc does not help, it only clears weakReferences. Softreferences stay around until the memory is needed so badly that an OOM would be triggered.
I've done a few hours of research at this point and could not find any consistent way to clear the old softReferences from the cache. The oom function inside the comment does work sometimes, but it's very hacky and does expand the jvm used memory to the max, so that is not ideal either.
The only other solution I found is [JVM TI](https://docs.oracle.com/javase/1.5.0/docs/guide/jvmti/jvmti.html#IterateOverReachableObjects) but using that from clojure would be very annoying and it's not part of every jvm.

TLDR: I think we are 95% of the way to the full solution and the final 5% are either a lot of work or have big drawbacks.

bbatsov commented:
Might also be a good idea to add a section in the README explaining some of the limitations for the fn-deps/refs. I've been meaning to document all the features of Orchard in the README, but I never got to doing so. We have to start somewhere and that's as good of a place as any.

vemv reviewed (commented):
Alright, looking good thanks!

Might give it one last look later today.

Could you update the changelog and PR description in the meantime?

Cyrik commented:
Done.

I also wrote a follow-up [post](https://lukas-domagala.de/blog/clojure-compiler-class-cache.html) on why the class cache is hard to clear and linked that.

vemv commented on src/orchard/xref.clj:
```diff
@@ -13,21 +15,55 @@
+  `v` can be a function value, a var or a symbol.
+   If a function was defined multiple times, old lambda deps will
+   be returned.
+   This does not return functions marked with meta :inline like +
```
```suggestion
   This does not return functions marked with meta :inline like `+`
```

vemv commented on src/orchard/xref.clj:
```diff
@@ -13,21 +15,55 @@
+      ;; this uses the implementation detail that the clojure compiler always
+      ;; prefixes names of lambdas with the name of its surrounding function class
+      (into #{} (comp (filter (fn [[k _v]] (clojure.string/includes? k f-class-name)))
+                      (map (fn [[_k v]] (.get ^java.lang.ref.Reference v)))
```
Better if this `v` doesn't shadow the surrounding `v`

vemv commented on src/orchard/xref.clj:
```diff
@@ -13,21 +15,55 @@
+    (.get classCache* clojure.lang.DynamicClassLoader)))
+
+(defn fn-deps
+  "Returns a set with all the functions invoked by `v` and its lambdas.
```
Please use a wording clearer than "its lambdas"

vemv reviewed (commented):
Almost there 👍 ty!

Cyrik commented:
If you think this is relevant to orchard I'll add a `fn-transitive-deps` function to this PR. It would just be a recursive walk over the dependencies, so it might also be ok for users to write that themselfs.

bbatsov commented on test/orchard/xref_test.clj:
```diff
@@ -38,4 +41,6 @@
     (is (= (xref/fn-refs 'orchard.xref-test/dummy-fn) '()))
-    (is (contains? (into #{} (xref/fn-refs #'map)) #'orchard.xref-test/dummy-fn))))
+    (is (contains? (into #{} (xref/fn-refs #'map)) #'orchard.xref-test/dummy-fn)))
+  (testing "with a lambda"
```
Not sure if this test has any value, given that `times` now appears in all the other tests.

vemv replied:
> Not sure if this test has any value, given that times now appears in all the other tests.

Right, please craft a separate example such that it only a lambda dependency, so that it's super clear what we're testing.

i.e. it's not ideal to test many things at the same time.

As in a prior note, please avoid "lambda" terminology, in a deftest I'd suggesting a super clear wording so that we can maintain this effectively.

Finally, you can also rename `times` to something like `fn-dep` so that we emphasize what is being tested.

bbatsov commented on src/orchard/xref.clj:
```diff
@@ -45,8 +81,40 @@
+  (Math/min 1 2)
+  (def vars (q/vars {:ns-query {:project? true} :private? true}))
+
+  (map fn-deps vars))
```
The final newline seems to be missing.

bbatsov commented on src/orchard/xref.clj:
```diff
@@ -45,8 +81,40 @@
         deps-map (zipmap all-vars (map fn-deps all-vars))]
     (map first (filter (fn [[_k v]] (contains? v var)) deps-map))))
+
+(comment
```
Are those some functions you used for test purposes? Might be good idea to add a comment explaining this.

bbatsov commented:
Looks good to me overall. I even read the blog post - well done! I added a few small remarks and we are good to go. I guess we'll squash this to a single comment in the final version.

Cyrik commented:
I've added fn-transitive-deps, if you don't want it in here I'll throw it out again or put it in another PR.
Also added one more caveat to the readme. Didn't know this didn't work on AoT code, just found that out because of the transitive test.

vemv commented on test/orchard/xref_test.clj:
```diff
@@ -38,4 +44,12 @@
+  (testing "with a lambda"
+    (is (contains? (into #{} (xref/fn-refs #'times)) #'orchard.xref-test/dummy-fn))))
+
+(deftest fn-transitive-deps-test
+  (testing "basics"
+    (is (= (xref/fn-transitive-deps dummy-fn)
+           #{#'clojure.core/even? #'clojure.core/filter #'clojure.core/map
+             #'clojure.core/range #'orchard.xref-test/times #'orchard.xref-test/times*}))))
```
```suggestion
(deftest fn-transitive-deps-test
  (testing "basics"
    (let [expected #{#'clojure.core/even? #'clojure.core/filter #'clojure.core/map
                     #'clojure.core/range #'orchard.xref-test/times #'orchard.xref-test/times*}]
      (is (contains? expected #'orchard.xref-test/times*)
          "Specifically includes `#'times*`, which is a transitive dep of `#'dummy-fn` (via `#'times`)")
      (is (= expected
             (xref/fn-transitive-deps dummy-fn))))))
```

vemv reviewed (commented):
One final note from my side

vemv commented:
Cheers 🍻
